integration/conftest.py

```diff
@@ -150,133 +157,34 @@ def flog_binary():
-            join(gather_dir, 'gatherer.tac'),
-        ),
-        path=gather_dir,
-        env=environ,
-    )
-    pytest_twisted.blockon(twistd_protocol.magic_seen)
-
-    def cleanup():
-        _cleanup_tahoe_process(twistd_process, twistd_protocol.exited)
-
-        flog_file = mktemp('.flog_dump')
-        flog_protocol = _DumpOutputProtocol(open(flog_file, 'w'))
-        flog_dir = join(temp_dir, 'flog_gather')
-        flogs = [x for x in listdir(flog_dir) if x.endswith('.flog')]
-
-        print("Dumping {} flogtool logfiles to '{}'".format(len(flogs), flog_file))
-        reactor.spawnProcess(
-            flog_protocol,
-            flog_binary,
-            (
-                'flogtool', 'dump', join(temp_dir, 'flog_gather', flogs[0])
-            ),
-            env=environ,
-        )
-        print("Waiting for flogtool to complete")
-        try:
-            block_with_timeout(flog_protocol.done, reactor)
-        except ProcessTerminated as e:
-            print("flogtool exited unexpectedly: {}".format(str(e)))
-        print("Flogtool completed")
-
-    request.addfinalizer(cleanup)
-
-    with open(join(gather_dir, 'log_gatherer.furl'), 'r') as f:
-        furl = f.read().strip()
-    return furl
+    return fg
 
 
 @pytest.fixture(scope='session')
-@log_call(
-    action_type=u"integration:introducer",
-    include_args=["temp_dir", "flog_gatherer"],
-    include_result=False,
-)
-def introducer(reactor, temp_dir, flog_gatherer, request):
-    intro_dir = join(temp_dir, 'introducer')
-    print("making introducer", intro_dir)
-
-    if not exists(intro_dir):
-        mkdir(intro_dir)
-        done_proto = _ProcessExitedProtocol()
-        _tahoe_runner_optional_coverage(
-            done_proto,
-            reactor,
-            request,
-            (
-                'create-introducer',
-                '--listen=tcp',
-                '--hostname=localhost',
-                intro_dir,
-            ),
-        )
-        pytest_twisted.blockon(done_proto.done)
-
-    config = read_config(intro_dir, "tub.port")
-    config.set_config("node", "nickname", "introducer-tor")
-    config.set_config("node", "web.port", "4562")
-    config.set_config("node", "log_gatherer.furl", flog_gatherer)
-
-    # "tahoe run" is consistent across Linux/macOS/Windows, unlike the old
-    # "start" command.
-    protocol = _MagicTextProtocol('introducer running', "introducer")
-    transport = _tahoe_runner_optional_coverage(
-        protocol,
-        reactor,
-        request,
-        (
-            'run',
-            intro_dir,
-        ),
+@log_call(action_type=u"integration:grid", include_args=[])
+def grid(reactor, request, temp_dir, flog_gatherer, port_allocator):
+    # XXX think: this creates an "empty" grid (introducer, no nodes);
+    # do we want to ensure it has some minimum storage-nodes at least?
+    # (that is, semantically does it make sense that 'a grid' is
+    # essentially empty, or not?)
```

Comment:
I don't see any way to _rule out_ the possibility that a client can't reach any other nodes ... so since "empty grid" is a real world thing, I don't see any reason not to admit it into the tests.

Reply:
I'll at least remove the comment.

The practical aspect (which I apparently didn't write down) is basically do we want to have tests declare `grid` _and_ `storage_servers` dependencies/fixtures or document/imply that `grid` includes .. something?

It's definitely more flexible the way it is. I think I wrote that down when I discovered that declaring `grid` wasn't usually sufficient for tests (because they typically want to "do stuff" so need storage-servers).

Tests using the utilities directly can always get "empty grid" this is moreso about the approach for the `grid` fixture itself

Reply:
Hmm, after writing the above, I note that tests that want usable storage-servers actually _only_ need to declare `storage_servers` and forget about `grid`...so maybe that is more clear in the end.